docs/USER_NAMES.md

```diff
@@ -0,0 +1,164 @@
+to register a user or group in the system, for example a system user/group
+using
+[`systemd-sysusers.service`](https://www.freedesktop.org/software/systemd/man/systemd-sysusers.html)
+or a regular user with
+[`systemd-homed.service`](https://www.freedesktop.org/software/systemd/man/systemd-homed.html).
+
+In strict mode, only uppercase and lowercase characters are allowed, as well as
+digits, underscores and hyphens. The first character may not be a digit or
+hyphen. A size limit is enforced: the minimum of `sysconf(_SC_LOGIN_NAME_MAX)`
+(typically 256 on Linux; rationale: this is how POSIX suggests to detect the
+limit), `UT_NAMESIZE-1` (typically 31 on Linux; rationale: names longer than
+this cannot correctly appear in `utmp`/`wtmp` and create ambiguity with login
+accounting) and `FILENAME_MAX` (4096 on Linux; rationale: user names typically
+appear in directory names, i.e. the home directory), thus MIN(256, 31, 4096) =
+31.
+
+Note that these rules are both more strict and more relaxed than all of the
+rules enforced by other systems listed above. A user/group name conforming to
+systemd's strict rules will not necessarily pass a test by the rules enforced
+by these other subsystems.
+
+Written as regular expression the above is: `^[a-zA-Z_][a-zA-Z0-9_-]{0,30}$`
+
+### Relaxed mode
+
+Relaxed user/group name syntax is enforced whenever a systemd component accepts
+and makes use of user/group names registered by other (non-systemd)
+components of the system, for example in
+[`systemd-logind.service`](https://www.freedesktop.org/software/systemd/man/systemd-logind.html).
+
+Relaxed syntax is also enforced by the `User=` setting in service unit files,
+i.e. for system services used for running services. Since these users may be
+registered by a variety of tools relaxed mode is used, but since the primary
+purpose of these users is to run a system service and thus a job for systemd a
+warning is shown if the specified user name does not qualify by the strict
+rules above.
+
+* No embedded NUL bytes (rationale: handling in C must be possible and
+  straight-forward)
+
+* No names consisting fully of digits (rationale: avoid confusion with numeric
+  UID/GID specifications)
+
+* Similar, no names consisting of an initial hyphen and otherwise entirely made
+  up of digits (rationale: avoid confusion with negative, numeric UID/GID
+  specifications, e.g. `-1`)
+
+* No strings that do not qualify as valid UTF-8 (rationale: we want to be able
+  to embed these strings in JSON, with permits only valid UTF-8 in its
+  strings)
+
+* No control characters (i.e. characters in ASCII range 1…31; rationale: they
+  tend to have special meaning when output on a terminal in other contexts,
+  moreover the newline character — as a specific control character — is used as
+  record separator in `/etc/passwd`, and hence it's crucial to avoid
+  ambiguities here)
+
+* No colon characters (rationale: it is used as field separator in `/etc/passwd`)
```

Comment:
I get the feeling that the ancient Unix tools need some modernization. Perhaps there could be a new format without these restrictions which could be handled by systemd-userdb. Compatibility to traditional passwd format could be done by encoding the real names (or just UIDs) with something like base64 but never presenting those to users.

Reply:
The truth s that we could completely replace and obsolete the /etc/passwd file.
It is not strictly required as all relevant software uses the nsswitch facility to access this data.
however there may be some early boot/static tools that may be still using fgetpwent() those would need to be change to access whatever replacement for /etc/passwd (or obsoleted and replaced).

Reply:
It has annoyed me greatly since forever that setuid/setgid helpers (passwd, chsh, chfn, unix_chkpwd) are needed because of compatibility to ancient formats (GECOS? It's 2020). If the user data would be directly writable by the user (/etc/passwd.d/$UID, with mode 0660, GID $GID, or more modernly with ACLs, or even have all data managed by systemd-userdb), this would not be needed. Also I think nobody loves the ugly NSS libraries supplied by libc.

Reply:
Well, getting rid of /etc/passwd entirely (i.e. not just as file, but as format) is a very difficult thing, because it lives on in "getent passwd"'s output and suchlike. It's not something I would really bother with. I mean, why even? The ability of using ":" in usernames sounds like an awfully weak reason to me, to be worth the work for doing that...

Anyway, this PR is supposed to be something we can do today, anything like getting rid of /etc/passwd as a concept/format sounds to me to be work for another time...

Reply:
Agreed, I was just mentioning that it is possible, not endorising doing it, it would cause unnecessary churn for basically no gain.